wrappers/core18.go

```diff
@@ -141,16 +141,49 @@ type AddSnapdSnapServicesOptions struct {
 	Preseeding bool
 }
 
+type SnapdRestart struct {
+	Sysd systemd.Systemd
+}
+
+func (r *SnapdRestart) Restart() error {
+	if err := r.Sysd.StartNoBlock([]string{"snapd.apparmor.service"}); err != nil {
+		return err
+	}
+
+	// and finally start snapd.service (it will stop by itself and gets
+	// started by systemd then)
+	// Because of the file lock held on the snapstate by the Overlord, the new
+	// snapd will block there until we release it. For this reason, we cannot
+	// start the unit in blocking mode.
+	// TODO: move/share this responsibility with daemon so that we can make the
+	// start blocking again
+	if err := r.Sysd.StartNoBlock([]string{"snapd.service"}); err != nil {
```

Comment:
if we error here or before, is not clear that the snapd or snapd-failure will come into play at all, so we need to do something, no?

Reply:
So I have been trying to make this one fail. But I am not sure we can. We get more catastrophic failure. But I think it is already the case in master, because we get killed before we can even restart.

What I want is a spread test that exercise the behavior we want. Or at least a manual test. But it is not obvious how to make this fail.

But let's for some reason systemd is busy and does not not answer and `systemctl` just fails. Then yes, it theory, we have not failure running again. But this problem can also occur when `systemctl` tells us "all good", but systemd later for some reason (maybe broken dependency), does not start snapd or snapd-failure. I think both case of error or no error, we should make sure the new snapd get restarted before we exit. And if we timeout, we have to rollback. It is not safe to assume we can restart ourself until we do. Symlink or not, because the `snapd.service might be completely broken.